plugins/destination/batchingwriter/batchingwriter.go

```diff
@@ -0,0 +1,268 @@
+package batchingwriter
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/apache/arrow/go/v13/arrow"
+	"github.com/apache/arrow/go/v13/arrow/util"
+	"github.com/cloudquery/plugin-pb-go/specs"
+	"github.com/cloudquery/plugin-sdk/v3/plugins/destination"
+	"github.com/cloudquery/plugin-sdk/v3/schema"
+	"github.com/rs/zerolog"
+)
+
+type Batching struct {
+	underlyingWriter destination.ManagedWriter
+	logger           zerolog.Logger
+
+	metrics     map[string]*destination.Metrics
+	metricsLock *sync.RWMutex
+
+	workers     map[string]*worker
+	workersLock *sync.Mutex
+
+	dedupPK        bool
+	batchSize      int64
+	batchSizeBytes int64
+	batchTimeout   time.Duration
+
+	// underlyingOCW is set if the given underlyingWriter implements OpenCloseWriter
+	underlyingOCW OpenCloseWriter
+}
+
+// OpenCloseWriter is an optional interface that can be implemented by a Client which already implements destination.ManagedWriter.
+type OpenCloseWriter interface {
+	OpenTable(ctx context.Context, sourceSpec specs.Source, table *schema.Table, syncTime time.Time) error
+	CloseTable(ctx context.Context, sourceSpec specs.Source, table *schema.Table, syncTime time.Time) error
+	destination.ManagedWriter
+}
+
+func New(opts ...Option) destination.BatchingWriterFuncFunc {
```

Comment:
`destination.BatchingWriterFuncFunc` is odd

Reply:
@candiduslynx this is due to the initialization order... there's a chicken-egg problem :)